File: StockM.Web/src/services/riskManager.ts
import { StockSignal, RiskParameters, RiskAssessment } from '../types';

export function evaluateRisk(signal: StockSignal, params: RiskParameters): RiskAssessment {
  const entry = signal.entryPrice;
  const sl = signal.stopLoss;
  const tp = signal.takeProfit;

  const riskPerShare = entry - sl;
  const rewardPerShare = tp - entry;

  const maxRiskAmount = params.portfolioValue * params.maxPositionSizePct / 100;
  let shareCount = riskPerShare > 0 ? Math.floor(maxRiskAmount / riskPerShare) : 0;

  let volAdj = false;
  if (signal.indicators.atr > entry * 0.03) {
    shareCount = Math.floor(shareCount * (1 - params.volatilityReductionPct / 100));
    volAdj = true;
  }

  const maxLoss = round2(shareCount * riskPerShare);
  const maxGain = round2(shareCount * rewardPerShare);
  const rrRatio = maxLoss > 0 ? round2(maxGain / maxLoss) : 0;

  return {
    stopLossPrice: sl,
    takeProfitPrice: tp,
    recommendedPositionPct: signal.positionSizePct,
    shareCount,
    maxLoss,
    maxGain,
    riskRewardRatio: rrRatio,
    volatilityAdjusted: volAdj,
  };
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

File: StockM.Desktop/Services/RiskManager.cs
using StockM.Desktop.Models;

namespace StockM.Desktop.Services;

public class RiskManager
{
    public RiskAssessment Evaluate(StockSignal signal, RiskParameters parameters)
    {
        var entry = signal.EntryPrice;
        var sl = signal.StopLoss;
        var tp = signal.TakeProfit;

        var riskPerShare = entry - sl;
        var rewardPerShare = tp - entry;

        // Position sizing: risk no more than X% of portfolio on one trade
        var maxRiskAmount = parameters.PortfolioValue * (decimal)parameters.MaxPositionSizePct / 100m;
        int shareCount = riskPerShare > 0
            ? (int)(maxRiskAmount / riskPerShare)
            : 0;

        // Volatility adjustment
        bool volAdj = false;
        if (signal.Indicators.ATR > entry * 0.03m)
        {
            shareCount = (int)(shareCount * (1 - parameters.VolatilityReductionPct / 100));
            volAdj = true;
        }

        var maxLoss = shareCount * riskPerShare;
        var maxGain = shareCount * rewardPerShare;
        var rrRatio = maxLoss > 0 ? (double)(maxGain / maxLoss) : 0;

        return new RiskAssessment
        {
            StopLossPrice = sl,
            TakeProfitPrice = tp,
            RecommendedPositionPct = signal.PositionSizePct,
            ShareCount = shareCount,
            MaxLoss = Math.Round(maxLoss, 2),
            MaxGain = Math.Round(maxGain, 2),
            RiskRewardRatio = Math.Round(rrRatio, 2),
            VolatilityAdjusted = volAdj
        };
    }
}
